### PR TITLE
[InstCombine] Handle more even/odd math functions

### LINKED ISSUE #77220: Handle `f(-x)=f(x)` for even functions and `f(-x)=-f(x)` for odd functions
Currently, Clang handles this for trigonometric functions (`sin`, `cos`, and `tan`). 
https://godbolt.org/z/eharx1Gn4

However, for example, Clang doesn't currently handle this for the error function (`erf`).
https://godbolt.org/z/33W9Eharo
```c
#include<math.h>

double test_erf1(double x)
{
   return -erf(-x);
}

double test_erf2(double x)
{
   return erf(x);
}
``` 

There are many other functions we have to check, but I think the following GCC test is helpful.
https://github.com/gcc-mirror/gcc/blob/master/gcc/testsuite/gcc.dg/torture/builtin-symmetric-1.c

### PR DESCRIPTION
**Pull request by agentcooper**
At the moment PR adds support only for `erf` function.

Does it make sense to create a list of known even/odd functions and call the newly introduced `LibCallSimplifier::optimizeSymmetric` at the point where `optimizeTrigReflections` is now called?

Fixes #77220.

### CODE REVIEW

File: llvm/include/llvm/Analysis/TargetLibraryInfo.def
Comment (dtcxzyw) [inline llvm/include/llvm/Analysis/TargetLibraryInfo.def:1845]:
```suggestion
/// float erff(float x);
```

File: llvm/include/llvm/Analysis/TargetLibraryInfo.def
Comment (dtcxzyw) [inline llvm/include/llvm/Analysis/TargetLibraryInfo.def:1850]:
```suggestion
/// long double erfl(long double x);
```

File: llvm/include/llvm/Analysis/TargetLibraryInfo.def
Comment (dtcxzyw) [inline llvm/include/llvm/Analysis/TargetLibraryInfo.def:1840]: Please insert these funcs in the lexicographic order.

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2803]: Please add a test for the multi-use case.

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2804]: Please add a test for FMF preserving.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (agentcooper) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2789]: Could you please point me to a similar test?

**Comment by agentcooper** [comment]
@pogo59 Could you please help with updating `llvm/test/tools/llvm-tli-checker/ps4-tli-check.yaml`? It is unclear to me how it should be updated when adding new `math.h` functions.

**Comment by Krishna-13-cyber** [comment]
> @pogo59 Could you please help with updating `llvm/test/tools/llvm-tli-checker/ps4-tli-check.yaml`? It is unclear to me how it should be updated when adding new `math.h` functions.

Hi @agentcooper,
You can just set the added functions as unavailable for PS4/PS5 in TargetLibraryInfo.cpp which will help you pass all the checks.
```
    TLI.setUnavailable(LibFunc_erf);
    TLI.setUnavailable(LibFunc_erff);
    TLI.setUnavailable(LibFunc_erfl);
```

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (agentcooper) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2790]: Done 👍

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (agentcooper) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2794]: Done. I've added FMF flags (`reassoc`) and general flags (`tail`) to the call site in the original test.

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:1925]: I think `f(fabs(x)) --> f(x)` and `f(copysign(x, y)) --> f(x)` also hold for even functions.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thanks!
Please wait for additional approval from other reviewers.

**Comment by pogo59** [comment]
> You can just set the added functions as unavailable for PS4/PS5 in TargetLibraryInfo.cpp which will help you pass all the checks.

That will work for now. I'll add an internal task to handle these properly on our side. Thanks!

**Comment by agentcooper** [comment]
@arsenm @jcranmer-intel Could you please take a look?

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (jcranmer-intel) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:1929]: You can wrap the `fabs` and `copysign` stuff into the symmetric call. e.g., note that `f(fabs(x))` is `f(x)` if `f` is even and `fabs(f(x))` if `f` is odd. That would eliminate this method entirely.

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (jcranmer-intel) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2800]: On a side note, I wonder if commuting the sign with an odd function is an improvement in practice. For even functions, we're clearly getting rid of an instruction, which is almost always a win. But for odd functions, it's just moving the instruction, and it's not offhand clear to me if peephole opportunities tend to be better with it being before or after the libm function.

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (agentcooper) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp]: Done ✅

I've decided to mimic the existing functionality, so do it only for the even case (where it has a clear win).

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (agentcooper) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2772]: I guess one option is to look only for cases like `-f(-x)` where `f` is odd, so it can be replaced with just `f(x)`.

**Review by jcranmer-intel — APPROVED** [review]

**Comment by agentcooper** [comment]
@jcranmer-intel @dtcxzyw Could you please merge this for me? I don't have commit access. The Linux build is green on CI and failure in the Windows build seems to be unrelated to this PR.

**Comment by dtcxzyw** [comment]
Thanks for your contribution!

**Pull request merged**
